add validator whitelist

MoonShiesty opened the pull request:
# Goals of PR

Core changes:

Add validator whitelist to prevent other actors from calling state-modifying functions.

I've done some basic testing, if this draft looks good, I'll add checks to validator functions and test.

chalkandpaste commented on contracts/src/IRollup.sol:
```diff
@@ -220,6 +228,21 @@ interface IRollup {
     function setBaseStakeAmount(uint256 newAmount) external;
 
+    /**
+     * @notice Set the addresses of the validator whitelist
```
Can you write a better notice? I'm thinking this interface can be a little unintuitive, in that if an address previously whitelisted, but doesn't have a state update supplied to this function, they will remain whitelisted (that is, they have to be specifically unwhitelisted). Maybe a better function name might clear this up as well? `updateValidators`.

MoonShiesty replied:
pushed the change

MoonShiesty commented:
@chalkandpaste @siosw ready for review: added testing and renamed things for clarity.

I added the `whitelistOnly` modifier to the following validator functions which could be misused in testnet:

- `stake`
- `unstake`
- `removeStake`
- `advanceStake`
- `withdraw`
- `createAssertion`
- `challengeAssertion`
- `confirmFirstUnresolvedAssertion`
- `rejectFirstUnresolvedAssertion`

The way the whitelist is designed if we remove an already staked validator from the whitelist, that validator will not be able to `unstake` itself - another whitelisted addresses will have to call `removeStake` for them. They also won't be able to `withdraw` funds won in challenges. If this is an issue, we should remove the modifier for `unstake` and `withdraw`.

`unstake`/`advanceStake`/`createAssertion` already check that the caller is staked, but adding `whitelistOnly` allows us to disable all validator functionality for addresses that are already staked. This might be check depending on how we plan to use the whitelist.

chalkandpaste commented:
Re: `unstake` and `removeStake`... This could be solved by removing the stake when they are removed from the whitelist. Alternatively, but maybe not so clean, is to *not* use the modifier on `unstake` given it's restriction to the caller. It feels like maybe there's a third way... but nothing's coming to me.

Re: the second bit... this kinda indicates to me it might best to remove the stake when the validator is removed from the whitelist.

chalkandpaste commented:
Otherwise, everything looks good.

MoonShiesty commented:
> Re: `unstake` and `removeStake`... This could be solved by removing the stake when they are removed from the whitelist. Alternatively, but maybe not so clean, is to _not_ use the modifier on `unstake` given it's restriction to the caller. It feels like maybe there's a third way... but nothing's coming to me.
> 
> Re: the second bit... this kinda indicates to me it might best to remove the stake when the validator is removed from the whitelist.

This approach would be tricky because we don't want to remove the stake of a validator that is staked on an unconfirmed assertion.

I'm going to change the implementation so we don't `onlyWhitelist` `withdraw` - `withdraw` only affects the calling validator's state. This way un-whitelisted validators can remove earnings. 

I think we should still `onlyWhitelist` `unstake` because `removeStake` still allows whitelisted validators to return un-whitelisted stake. But I could go either way, as long we document.

I'm also going to document which functions might be whitelisted.

siosw commented on contracts/test/Rollup.t.sol:
```diff
@@ -224,6 +228,130 @@ contract RollupTest is RollupBaseSetup {
+    // Admin
+    ///////////////
+
+    function test_initialize_validatorWhitelist_works() external {
```
very much a nit but i think the convention is to name it `_succeeds()`

siosw commented on contracts/src/Rollup.sol:
```diff
@@ -210,7 +226,28 @@ contract Rollup is RollupBase {
+    function updateValidatorWhitelist(address[] calldata validatorWhitelistUpdate, bool[] calldata vals)
+        external
+        override
+        onlyOwner
```
should there be a way to for validators to revoke their own permissions? 
might be useful in the case of compromised keys

MoonShiesty replied:
good idea

siosw commented on contracts/src/Rollup.sol:
```diff
@@ -210,7 +226,28 @@ contract Rollup is RollupBase {
 
     /// @inheritdoc IRollup
-    function stake() external payable override {
+    function updateValidatorWhitelist(address[] calldata validatorWhitelistUpdate, bool[] calldata vals)
```
my personal preference would be to split this into 2 functions like
```
function whitelistValidators(address[] calldata validatorsToAdd)
function unwhitelistValidators(address[] calldata validatorsToRemove)
```

MoonShiesty commented on contracts/package.json:
```diff
@@ -42,8 +42,8 @@
   "dependencies": {
-    "@openzeppelin/contracts": "^4.8.3",
-    "@openzeppelin/contracts-upgradeable": "^4.8.3",
+    "@openzeppelin/contracts": "^4.9.3",
```
Updated to newer openzepplin with the AccessControlDefaultAdminRulesUpgradeable extension.

MoonShiesty commented on contracts/src/Rollup.sol:
```diff
@@ -210,7 +229,53 @@ contract Rollup is RollupBase {
-    function stake() external payable override {
+    function addValidatorsToWhitelist(address[] calldata validators) external override onlyRole(DEFAULT_ADMIN_ROLE) {
+        for (uint256 i = 0; i < validators.length; i++) {
+            if (hasRole(VALIDATOR_WHITELIST_ROLE, validators[i])) {
```
We can simplify this check in the next OpenZeppelin version which returns a bool for grant/RevokeRole() to indicate whether the call succeeds.

MoonShiesty commented on contracts/test/Rollup.t.sol:
```diff
@@ -224,11 +230,212 @@ contract RollupTest is RollupBaseSetup {
+        validators[0] = alice;
+
+        // Adding a validator to whitelist as non-admin should fail
+        vm.expectRevert(
```
We can simplify this test with a selector in the next OpenZeppelin version which defines custom error.

MoonShiesty commented:
Ready for review. I changed the validator whitelist implementation to use role-based access control. There are currently two roles, `VALIDATOR_WHITELIST_ROLE` and `DEFAULT_ADMIN_ROLE`.

Role are managed by the admin role, which is initialized to the proxy deployer. The admin role is governed by `AccessControlDefaultAdminRules` which implements the following mitigations: 

- Only one account holds the `DEFAULT_ADMIN_ROLE` since deployment until it’s potentially renounced.

- Enforces a 2-step process to transfer the `DEFAULT_ADMIN_ROLE` to another account.

- Enforces a configurable delay between the two steps, with the ability to cancel before the transfer is accepted.

- The delay can be changed by scheduling, see [changeDefaultAdminDelay](https://docs.openzeppelin.com/contracts/4.x/api/access#AccessControlDefaultAdminRules-changeDefaultAdminDelay-uint48-).
- It is not possible to use another role to manage the `DEFAULT_ADMIN_ROLE`.

I set the initial delay to change the admin role to 3 days.

Also added more testing and documentation.

LEAFERx commented on contracts/deploy/l1/02_deploy_Rollup.ts:
```diff
@@ -46,6 +46,7 @@ const func: DeployFunction = async function (hre: HardhatRuntimeEnvironment) {
     0, // uint256 _initialAssertionID
     0, // uint256 _initialInboxSize
     initialVMHash, // bytes32 _initialVMhash
+    [validator], // address[] calldata _validators
```
sequencer is also a validator

LEAFERx commented on contracts/src/Rollup.sol:
```diff
@@ -61,7 +67,6 @@ abstract contract RollupBase is
 
     function __RollupBase_init() internal onlyInitializing {
-        __Ownable_init();
         __UUPSUpgradeable_init();
```
do you need to init `AccessControlDefaultAdminRulesUpgradeable`?

LEAFERx commented on contracts/src/Rollup.sol:
```diff
@@ -115,6 +128,12 @@ contract Rollup is RollupBase {
         lastCreatedAssertionID = _initialAssertionID;
 
+        // Initialize role based access control
+        __AccessControlDefaultAdminRules_init(3 days, msg.sender);
```
maybe move to `__RollupBase_init()` so we have a consistent place to initialize these contracts?

MoonShiesty commented:
I've left `unstake` as a non-whitelisted function, because there's no way for a validator to misuse the `unstake` function - they would only be wasting gas. Otherwise if we remove a validator, with active stake from the whitelist, they won't be able to withdraw their stake. They would need another a whitelisted validator to call `removeStake` on their behalf.

I think its better to design the contracts so that validators can always trust that their stake is available to withdraw after the staked assertion is confirmed.

MoonShiesty commented on contracts/src/Rollup.sol:
```diff
@@ -126,20 +144,18 @@ contract Rollup is RollupBase {
         );
         emit AssertionCreated(lastCreatedAssertionID, msg.sender, _initialVMhash);
-
-        __RollupBase_init();
```
I moved `__RollupBase_init` to the top of the function since we need to initialize the admin role, before we can add roles and initialize the initial validator whitelist.

MoonShiesty commented:
@LEAFERx thanks for the review, added the requested changes

MoonShiesty commented on contracts/src/Rollup.sol:
```diff
@@ -210,7 +229,53 @@ contract Rollup is RollupBase {
+    {
+        for (uint256 i = 0; i < validators.length; i++) {
+            if (!hasRole(VALIDATOR_WHITELIST_ROLE, validators[i])) {
+                revert NoRoleToRevoke();
```
good point, especially since you'd need to look through events to see which accounts that have/have not been granted a role

MoonShiesty replied:
done

MoonShiesty commented:
@chalkandpaste thanks for the review, I've changed `addValidator/removeValidator` to use a single address, rather than a list. Let me know if it looks good.

I also squashed down to a single commit.
